File: 2300-construct-string-with-repeat-limit/2300-construct-string-with-repeat-limit.cpp
class Solution {
public:
    string repeatLimitedString(string s, int repeatLimit) {
        vector<int> freq(26, 0);

        for(auto &ch: s)
        {
            freq[ch - 'a']++;
        }

        priority_queue< pair<char, int>> pq;
        for(int i=25; i>=0; i--)
        {
            if(freq[i] == 0)
                continue;
            
            pq.push({'a'+i, freq[i]});
        }

        string ans = "";
        while(!pq.empty())
        {
            auto [ch1, cnt1] = pq.top();
            pq.pop();

            int addCnt = min(repeatLimit, cnt1);
            ans.append(addCnt, ch1);

            cnt1 -= addCnt;

            if(cnt1 > 0 && pq.empty() == false)
            {
                auto [ch2, cnt2] = pq.top();
                pq.pop();

                ans.push_back(ch2);
                cnt2--;

                pq.push({ch1, cnt1});
                if(cnt2 > 0)
                    pq.push({ch2, cnt2});
            }
        }

        return ans;
    }
};

File: 2300-construct-string-with-repeat-limit/2300-construct-string-with-repeat-limit.py
import heapq

class Solution:
    def repeatLimitedString(self, s: str, repeatLimit: int) -> str:
        freq = [0] * 26
        for ch in s:
            freq[ord(ch) - ord('a')] += 1

        pq = []
        for i in range(25, -1, -1):
            if freq[i] > 0:
                heapq.heappush(pq, (-i, freq[i]))

        ans = []
        while pq:
            i1, cnt1 = heapq.heappop(pq)
            ch1 = chr(-i1 + ord('a'))
            addCount = min(repeatLimit, cnt1)
            ans.append(ch1 * addCount)
            cnt1 -= addCount

            if cnt1 > 0 and pq:
                i2, cnt2 = heapq.heappop(pq)
                ch2 = chr(-i2 + ord('a'))
                ans.append(ch2)
                cnt2 -= 1

                heapq.heappush(pq, (i1, cnt1))
                if cnt2 > 0:
                    heapq.heappush(pq, (i2, cnt2))

        return ''.join(ans)
